internal/olm/client/status.go

```diff
@@ -71,13 +87,74 @@ func (c Client) GetObjectsStatus(ctx context.Context, objs ...runtime.Object) St
 	return Status{Resources: rss}
 }
 
-func (s Status) HasExistingResources() bool {
+// HasInstalledResources only returns true if at least one resource in s
+// was returned successfully by the API server. A resource error status
+// containing any error except "not found", or "no kind match" errors
+// for Custom Resources, will result in HasInstalledResources returning
+// false and the error.
+func (s Status) HasInstalledResources() (bool, error) {
+	crdKindSet, err := s.getCRDKindSet()
+	if err != nil {
+		return false, fmt.Errorf("error getting set of CRD kinds in resources: %w", err)
+	}
 	for _, r := range s.Resources {
 		if r.Resource != nil {
-			return true
+			return true, nil
+		} else if r.Error != nil && !apierrors.IsNotFound(r.Error) {
+			nkmerr := &meta.NoKindMatchError{}
+			if !errors.As(r.Error, &nkmerr) || !crdKindSet.Has(r.GVK.Kind) {
```

Comment:
Can you confirm my understanding of this conditional?

> `!errors.As(r.Error, &nkmerr)`

We get a NoKindMatchError when we request a CR whose CRD is not present in the cluster? And if the CRD is not present, then it isn't possible for the CR to be present, so that error is effectively the same as a not found error.

If the error is not a NoKindMatchError, we return the error.

> !crdKindSet.Has(r.GVK.Kind)

We only get to this check if we have a NoKindMatchError. The `crdKindSet` contains the Kinds of the CRDs that we fetched, but they aren't necessarily applied to the cluster.

So if we have a NoKindMatchError and the resource we fetched from the OLM installation manifest (which must be a CR since we have a NoKindMatchError) is not a resource defined by the set of CRDs that we fetched, then we return the NoKindMatchError.

Is there a specific scenario you had in mind for this part of the check. It seems like this would only happen if the OLM resource manifest had some unknown CR in it.

It might help to add a few comments and/or break up the conditionals to make the logic easier to understand.

Reply:
Your understanding is correct. I'll add an explanatory comment above the conditional.

>Is there a specific scenario you had in mind for this part of the check. It seems like this would only happen if the OLM resource manifest had some unknown CR in it.

I'm thinking of a more arbitrary resource input scenario. We should not assume that all CR kinds in `Status` resources are associated with present CRD's in that case. This may potentially crop up when using `Status` for deploying operators (see #1912) since we're taking an arbitrary set of manifests the user has defined and deploying them. I don't expect CR's to necessarily be present in that set given how operator bundles are defined; nonetheless it would be nice to future-proof this.